File: Javascript/Array/E-20. Valid Parentheses.ts
function isValid(s: string): boolean {
  const parenthesesMap = {
    ']': '[',
    ')': '(',
    '}': '{',
  };
  const stack: string[] = [];
  for (const char of s) {
    /** if is not close brackets  */
    if (!(char in parenthesesMap)) {
      stack.push(char);
    } else {
      const openBracket = stack.pop();
      if (parenthesesMap[char] !== openBracket) return false;
    }
  }
  if (stack.length > 0) return false;
  else return true;
}

const testData = [
  { s: '()', ans: true },
  { s: '()[]{}', ans: true },
  { s: '(]', ans: false },
  { s: '[', ans: false },
];

for (const data of testData) {
  console.log(data.ans === isValid(data.s));
}

File: Python/Array/E-20. Valid Parentheses.py
class Solution:
    @classmethod
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False

        bracketMap = {
            "]": "[",
            "}": "{",
            ")": "("
        }
        stack = []
        for char in s:
            # open bracket
            if char not in bracketMap:
                stack.append(char)
            # close bracket
            else:
                open_bracket = stack.pop() if len(stack) > 0 else None
                if (open_bracket != bracketMap[char]):
                    return False

        return False if len(stack) > 0 else True


test_data = [
    {"s": "()", "ans": True},
    {"s": "()[]{}", "ans": True},
    {"s": "(]", "ans": False},
    {"s": "[", "ans": False},
    {"s": "]", "ans": False},
    {"s": "))", "ans": False}
]

for data in test_data:
    print(data["ans"] == Solution.isValid(data["s"]))
